src/app-layer-htp-range.h

```diff
@@ -0,0 +1,110 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+#ifndef __APP_LAYER_HTP_RANGE_H__
+#define __APP_LAYER_HTP_RANGE_H__
+
+#include "util-thash.h"
+
+void HttpRangeContainersInit(void);
+void HttpRangeContainersDestroy(void);
+uint32_t HttpRangeContainersTimeoutHash(struct timeval *ts);
+
+void *HttpRangeContainerUrlGet(const uint8_t *key, size_t keylen, struct timeval *ts);
+
+// linked list of ranges : buffer with offset
+typedef struct HttpRangeContainerBuffer {
+    /** red and black tree */
+    RB_ENTRY(HttpRangeContainerBuffer) rb;
+    /** allocated buffer */
+    uint8_t *buffer;
+    /** length of buffer */
+    uint64_t buflen;
+    /** the start of the range (offset relative to the absolute beginning of the file) */
+    uint64_t start;
+    /** offset of bytes written in buffer (relative to the start of the range) */
+    uint64_t offset;
+    /** number of gaped bytes */
+    uint64_t gap;
+} HttpRangeContainerBuffer;
+
+int HttpRangeContainerBufferCompare(HttpRangeContainerBuffer *a, HttpRangeContainerBuffer *b);
+
+RB_HEAD(HTTP_RANGES, HttpRangeContainerBuffer);
+RB_PROTOTYPE(HTTP_RANGES, HttpRangeContainerBuffer, rb, HttpRangeContainerBufferCompare);
+
+/** Item in hash table for a file in multiple ranges
+ * Thread-safety is ensured with the thread-safe hash table cf THashData
+ * The number of use is increased for each flow opening a new HttpRangeContainerBlock
+ * until it closes this HttpRangeContainerBlock
+ * The design goal is to have concurrency only on opening and closing a range request
+ * and have a lock-free data structure belonging to one Flow
+ * (see HttpRangeContainerBlock below)
+ * for every append in between (we suppose we have many appends per range request)
+ */
+typedef struct HttpRangeContainerFile {
+    /** key for hashtable */
+    uint8_t *key;
+    /** key length */
+    uint32_t len;
+    /** expire time in epoch */
+    uint32_t expire;
+    /** pointer to hashtable data, for locking and use count */
+    THashData *hdata;
+    /** total expected size of the file in ranges */
+    uint64_t totalsize;
+    /** size of the file after last sync */
+    uint64_t lastsize;
+    /** file container, with only one file */
+    FileContainer *files;
+    /** red and black tree list of ranges which came out of order */
+    struct HTTP_RANGES fragment_tree;
+    /** file flags */
+    uint16_t flags;
+    /** wether a HttpRangeContainerBlock is currently
+     owning the FileContainer in order to append to the file */
+    bool fileOwned;
+    /** error condition for this range. Its up to timeout handling to cleanup */
+    bool error;
+} HttpRangeContainerFile;
+
+/** A structure representing a single range request :
+ * either skipping, buffering, or appending
+ * As this belongs to a flow, appending data to it is ensured to be thread-safe
+ * Only one block per file has the pointer to the container
+ */
+typedef struct HttpRangeContainerBlock {
```

Comment:
after our conversation the other day I expected a `FileContainer` pointer to be here, where the flow handling the in-order chunk would move the `FileContainer` pointer into this block and return it to the flow on close.

Reply:
I had understood a `FileContainer` pointer or a boolean...
Switching to the `FileContainer` pointer then ;-)